source/simulator/solver.cc

```diff
@@ -853,6 +867,15 @@ namespace aspect
                                    solver_control_cheap,
                                    solver_control_expensive);
 
+        // We do not need the linearized_stokes_initial_guess vectory anymore, so
+        // (ab)use it for computing the initial residual when the Newton solver is used
+        // For the Newton solver we are interested in || F( x^(k-1) ) + F'( x^(k-1) )
+        // \partial x^(k-1) || to determine the correct linear tolerance, which is what
+        // we compute here.
+        if (assemble_newton_stokes_system)
+          initial_residual = stokes_block.residual (linearized_stokes_initial_guess,
+                                                    distributed_stokes_solution,
+                                                    system_rhs);
```

Comment:
This is the only part of the patch that I'm not actually clear about. You are computing the linear residual left after solving the linear system. If you were solving the linear system exactly, shouldn't this be zero? Or what am I missing?

Reply:
With the Eisenstat Walker method of determining the linear tolerance, the linear residual may actually be still relatively large. To compute the new linear tolerance we need to compute `(abs( || F( x_k )|| - || F( x_(k-1) ) + F'( x_(k-1) ) s_(k-1) || ) /|| F( x_(k-1) ) ||`, where `|| F( x_k )||` and `|| F( x_(k-1) ) ||` are the Newton rhs for this and the previous time step. So we only need the term: ` || F( x_(k-1) ) + F'( x_(k-1) ) s_(k-1) ||`, which we compute here.

Reply:
And s_(k-1) is the solution of the previous linear system? So the quantity you compute here should then be the last residual of the previous GMRES solve, right? If that's the case, I think we can get that out of the SolverControl object, but I first wanted to confirm.

Reply:
hmm, yes, that would then be better. Is it the last_value() which I need for that?

Reply:
Yes. I think you want to double-check first that these really give the same result plus or minus a few epsilon. But that should be it.

Reply:
So, I tested it and the problem is that there are two solver controllers: the cheap and the expensive one. If you use the wrong one (I tried outputing them with pcout) then you get `Floating point exception (core dumped)`. When the correct one is used, the correct controller is returned.

Reply:
Yes, you'll have to get the value from the correct `SolverControl` object. In particular, if the cheap solver succeeds, then you'll have to get that value. If the cheap solver fails, you'll have to get the value from the expensive solver.

You'll also have to be prepared to compute the value the way you do in the (original) patch if we happened to use the direct solver -- though, in that case, we could also just set the value to zero because we solved the system exactly.